# Atividade 1 — Exercícios com os comandos básicos de pandas para EDA


Responda as perguntas abaixo no notebook (cada resposta deve ter ao menos uma célula de código que produza o output e, logo abaixo, uma célula de texto com a sua interpretação).

## 1) Carregue o dataset e mostre as primeiras 10 linhas.

In [ ]:
import pandas as pd

df = pd.read_csv('emendas.csv')
df.head(10)

/tmp/ipykernel_12977/421898870.py:3: DtypeWarning: Columns (0,3,5) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('emendas.csv')


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"700000,00","0,00","0,00","0,00","690900,00","9100,00"
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"487500,00","0,00","0,00","0,00","487500,00","0,00"
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"100000,00","0,00","0,00","0,00","0,00","100000,00"
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","243750,00","0,00"
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"1658333,94","0,00","0,00","0,00","527200,00","1131133,94"
5,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPINAS DO SUL - RS,4303806,CAMPINAS DO SUL,4300000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","0,00","243750,00"
6,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,MORENO - PE,2609402,MORENO,2600000,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0000,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,"700000,00","0,00","0,00","0,00","700000,00","0,00"
7,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,NOVA FLORESTA - PB,2510105,NOVA FLORESTA,2500000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"278700,00","0,00","0,00","0,00","0,00","278700,00"
8,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTIAGO DO SUL - SC,4215695,SANTIAGO DO SUL,4200000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"300000,00","300000,00","300000,00","0,00","0,00","0,00"
9,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CIANORTE - PR,4105508,CIANORTE,4100000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"245850,00","0,00","0,00","0,00","0,00","245850,00"


Explique em uma frase o que você observa nas primeiras 10 linhas.

As primeiras 10 linhas mostram registros de emendas individuais do ano de 2014, detalhando a localização de aplicação do recurso (município/UF) e os valores destinados a diferentes ações orçamentárias.

## 2) Quantas linhas e colunas o dataset possui?

In [ ]:
df.shape

(71832, 28)

Explique o que esses números significam.

O dataset possui 71832 linhas (registros) e 28 colunas (variáveis/atributos), indicando o volume total de dados disponíveis para esta análise.

## 3) Liste os nomes das colunas e descreva brevemente o que você entende que cada coluna representa.

In [ ]:
df.columns

Index(['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda',
       'Código do Autor da Emenda', 'Nome do Autor da Emenda',
       'Número da emenda', 'Localidade de aplicação do recurso',
       'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região',
       'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção',
       'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação',
       'Código Plano Orçamentário', 'Nome Plano Orçamentário',
       'Valor Empenhado', 'Valor Liquidado', 'Valor Pago',
       'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados',
       'Valor Restos A Pagar Pagos'],
      dtype='object')

Escreva sua breve descrição das colunas, como se estivesse criando um dicionário de variáveis para que leitor entenda.

Código da Emenda: Identificador numérico único da emenda.

Ano da Emenda: Ano do exercício financeiro em que a emenda foi criada.

Tipo de Emenda: Classificação da origem (ex: Individual, de Bancada, de Relator).

Código do Autor da Emenda: ID numérico do parlamentar que indicou o recurso.

Nome do Autor da Emenda: Nome do Deputado ou Senador autor da emenda.

Número da emenda: Código sequencial da emenda no orçamento.

Localidade de aplicação do recurso: Descrição do destino final da verba.

Código Município IBGE / Município: Identificadores da cidade beneficiada.

Código UF IBGE / UF / Região: Identificadores do estado e região do país.

Código Função / Nome Função: Área de governo (ex: Saúde, Educação, Segurança).

Código Subfunção / Nome Subfunção: Segmento específico da área (ex: Atenção Básica).

Código Programa / Nome Programa: Nome do programa orçamentário do Governo.

Código Ação / Nome Ação: Atividade específica que receberá o investimento.

Código / Nome Plano Orçamentário: Detalhamento técnico da execução orçamentária.

Valor Empenhado: Valor que o governo se comprometeu a pagar (reservado).

Valor Liquidado: Valor onde o serviço/bem já foi comprovadamente entregue.

Valor Pago: Valor que já saiu dos cofres públicos para o beneficiário.

Valor Restos A Pagar Inscritos: Dívidas de anos anteriores reconhecidas.

Valor Restos A Pagar Cancelados: Valores de anos anteriores que não serão mais pagos.

Valor Restos A Pagar Pagos: Valores de anos anteriores que foram quitados agora..

## 4) Existem valores nulos? quantos?


In [ ]:
df.isnull().sum()

,0
Código da Emenda,0
Ano da Emenda,0
Tipo de Emenda,0
Código do Autor da Emenda,0
Nome do Autor da Emenda,0
Número da emenda,0
Localidade de aplicação do recurso,0
Código Município IBGE,0
Município,0
Código UF IBGE,0


## 5) Mostre os tipos das variáveis (se int, string, boolean, float...) e analise se alguma variável parece ter o tipo errado. Faça isso da maneira mais objetiva e otimizada possível.

In [ ]:
df.dtypes

,0
Código da Emenda,object
Ano da Emenda,int64
Tipo de Emenda,object
Código do Autor da Emenda,object
Nome do Autor da Emenda,object
Número da emenda,object
Localidade de aplicação do recurso,object
Código Município IBGE,object
Município,object
Código UF IBGE,int64


Se alguma coluna estiver com o tipo errado, escreva abaixo qual é e por que isso pode atrapalhar análises.

**Dica:** Se alguma variável tiver no tipo o nome "object", isso se refere a texto.

As colunas financeiras — Valor Empenhado, Valor Liquidado, Valor Pago e as colunas de Restos a Pagar — estão todas classificadas como object (texto).

Impossibilidade de cálculos: Como o Python entende que esses valores são "palavras", não podemos somar o total gasto, calcular a média de valores por região ou encontrar a emenda mais cara. Se tentarmos somar, o Python dará erro ou apenas juntará os textos.

## 6) Usando a coluna `Ano da Emenda` como base, descubra até que ano o dataset está atualizado (mostre as 8 primeiras linhas do df ordenado desc).

In [ ]:
df_ordenado = df.sort_values(by='Ano da Emenda', ascending=False)
df_ordenado.head(8)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
71783,202323680003,2023,Emenda Individual - Transferências com Finalid...,2368,FABIO RAMALHO,3,BELO HORIZONTE - MG,3106200,BELO HORIZONTE,3100000,...,2E90,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0000,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,"1104978,00","1104978,00","1104978,00","0,00","0,00","0,00"
71830,202331340003,2023,Emenda Individual - Transferências com Finalid...,3134,FAUSTO PINATO,3,MÚLTIPLO,Sem informação,Múltiplo,3500000,...,8535,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,0000,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,"1627846,00","482897,00","482897,00","7376,00","56517,00","1081056,00"
71812,202329790007,2023,Emenda Individual - Transferências com Finalid...,2979,FABIO REIS,7,Nacional,Sem informação,Sem informação,-1,...,2723,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...",0000,"POLICIAMENTO, FISCALIZACAO, COMBATE A CRIMINAL...","244381,00","0,00","0,00","0,00","0,00","244381,00"
71829,202331340002,2023,Emenda Individual - Transferências com Finalid...,3134,FAUSTO PINATO,2,MÚLTIPLO,Sem informação,Múltiplo,3500000,...,2E90,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0000,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,"6190995,00","6190995,00","6190995,00","0,00","0,00","0,00"
71828,202331340001,2023,Emenda Individual - Transferências com Finalid...,3134,FAUSTO PINATO,1,MÚLTIPLO,Sem informação,Múltiplo,3500000,...,2E89,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0000,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,"6606388,00","6606388,00","6606388,00","0,00","0,00","0,00"
71811,202329790006,2023,Emenda Individual - Transferências com Finalid...,2979,FABIO REIS,6,SERGIPE (UF),Sem informação,Sem informação,2800000,...,20Y6,"PESQUISA, DESENVOLVIMENTO E TRANSFERENCIA DE T...",0000,PESQUISA E DESENVOLVIMENTO DE TECNOLOGIAS PARA...,"170443,86","145415,76","145415,76","0,00","933,56","24094,54"
71810,202329790005,2023,Emenda Individual - Transferências com Finalid...,2979,FABIO REIS,5,Nacional,Sem informação,Sem informação,-1,...,2798,AQUISICAO E DISTRIBUICAO DE ALIMENTOS DA AGRIC...,0000,AQUISICAO E DISTRIBUICAO DE ALIMENTOS DA AGRIC...,"250000,00","0,00","0,00","0,00","0,00","250000,00"
71799,202323680020,2023,Emenda Individual - Transferências Especiais,2368,FABIO RAMALHO,20,MÚLTIPLO,Sem informação,Múltiplo,3100000,...,0EC2,TRANSFERENCIAS ESPECIAIS,0000,TRANSFERENCIAS ESPECIAIS,"1615740,00","1615740,00","1615740,00","0,00","0,00","0,00"


Explique: qual é o ano mais recente presente no dataset?

2023

## 7) execute um  `describe()` e explique uma conclusão simples que você consegue tirar a partir das estatísticas (se algumas colunas deveriam aparecer no describe mas não aparecem, indique o motivo que você acredita que esteja causando esse problema).

In [ ]:
df.describe()

,Ano da Emenda,Código UF IBGE
count,71832.000000,7.183200e+04
mean,2017.995796,2.981596e+06
std,2.664694,1.227537e+06
min,2014.000000,-1.400000e+01
25%,2016.000000,2.300000e+06
50%,2018.000000,3.100000e+06
75%,2020.000000,3.500000e+06
max,2023.000000,5.300000e+06


Conclusão Simples sobre os dados visíveis:
Período dos Dados: O dataset cobre emendas de 2014 (valor mínimo) até 2023 (valor máximo).

Média Temporal: A média dos anos é aproximadamente 2018, indicando que há um volume equilibrado de dados ao longo da última década.

Volume de Registros: O campo count mostra que temos 71.832 linhas com dados válidos nessas colunas.


Colunas que deveriam aparecer, mas não aparecem:
As colunas de Valor Empenhado, Valor Liquidado e Valor Pago são as mais importantes do dataset, mas ficaram de fora do resumo estatístico.


Motivo do problema:
Isso acontece porque essas colunas estão configuradas como o tipo object (texto), conforme vimos no teste anterior (df.dtypes).

## 8) Mostre as últimas 5 linhas do dataset e escreva se a última linha parece completa (sem muitos nulos) ou não.

In [ ]:
df.tail(5)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
71827,202328360003,2023,Emenda Individual - Transferências Especiais,2836,FABIO TRAD,3,MÚLTIPLO,Sem informação,Múltiplo,5000000,...,0EC2,TRANSFERENCIAS ESPECIAIS,0000,TRANSFERENCIAS ESPECIAIS,"14662883,00","14662883,00","14662883,00","0,00","0,00","0,00"
71828,202331340001,2023,Emenda Individual - Transferências com Finalid...,3134,FAUSTO PINATO,1,MÚLTIPLO,Sem informação,Múltiplo,3500000,...,2E89,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0000,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,"6606388,00","6606388,00","6606388,00","0,00","0,00","0,00"
71829,202331340002,2023,Emenda Individual - Transferências com Finalid...,3134,FAUSTO PINATO,2,MÚLTIPLO,Sem informação,Múltiplo,3500000,...,2E90,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0000,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,"6190995,00","6190995,00","6190995,00","0,00","0,00","0,00"
71830,202331340003,2023,Emenda Individual - Transferências com Finalid...,3134,FAUSTO PINATO,3,MÚLTIPLO,Sem informação,Múltiplo,3500000,...,8535,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,0000,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,"1627846,00","482897,00","482897,00","7376,00","56517,00","1081056,00"
71831,202331340004,2023,Emenda Individual - Transferências com Finalid...,3134,FAUSTO PINATO,4,MÚLTIPLO,Sem informação,Múltiplo,3500000,...,8581,ESTRUTURACAO DA REDE DE S,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Interprete a completude da última linha.

Ao executar o comando df.tail(5), observa-se que as últimas linhas do dataset apresentam informações sobre emendas do ano de 2023. A última linha parece estar incompleta, pois possue campos que não possuem dados preenchidos, com excesso de valores nulos (NaN) que indicariam um erro no carregamento do arquivo

## 9) existem valores duplicados no dataframe? Caso positivo, quantos?

In [ ]:
df.duplicated().sum()

np.int64(0)

Não existem valores duplicados no dataframe. Ao executar o comando df.duplicated().sum(), o retorno foi 0, o que confirma que cada linha do conjunto de dados representa um registro único. Isso garante que não haverá contagem dupla de valores financeiros ou distorções nas análises estatísticas por conta de repetições desnecessárias.